AssemblyName.ToString() returns an empty string for an empty `Name` property

BartoszKlonowski opened the pull request:
This pull request fixes #45420

It removes the *System.IO.FileLoadException* exception in case of calling `AssemblyName.ToString()` with empty `Name` property, making it returning the empty string instead.

---

The implementation of this fix is based on the `FullName` property which now returns the empty string *""* for both `null` and `string.Empty` cases.
(Note that `null` case is left intact).

Unit tests are now enhanced with the empty string for the `Name` property and `ToString()` method.
Also the `FullName` property test is reimplemented so from now on it does not expect it to throw the exception, but ensures that empty `Name` is equal to empty response.

BartoszKlonowski committed:
Return empty FullName of AssemblyName for empty Name

The emtpy Name of AssemblyName does not mean that there's a FileLoad
errror so there should not be the FileLoadException thrown for that
case.
This commit modifies the behavior of FullName getter, so the ToString
will return empty string "" when the Name is set to be empty.

NOTE: The null condition is left as is, even if the conditions of null
and empty are now combined.

BartoszKlonowski committed:
Cover the empty assembly name with unit tests

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs:
```diff
@@ -209,10 +209,8 @@ public string FullName
             get
             {
-                if (this.Name == null)
+                if (this.Name == null || this.Name == string.Empty)
```
```suggestion
                if (string.IsNullOrEmpty(this.Name))
```

jkotas committed:
Update src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs

jkotas reviewed (approved):
Thank you!

jkotas commented:
One of the tests is failing. Could you please take a look?
```
  Starting:    System.Reflection.Tests (parallel test collections = on, max threads = 2)
    System.Reflection.Tests.AssemblyNameTests.ToStringTest(name: "") [FAIL]
      System.ArgumentException : String cannot have zero length.
      Stack Trace:
        /_/src/coreclr/System.Private.CoreLib/src/System/Reflection/AssemblyName.CoreCLR.cs(20,0): at System.Reflection.AssemblyName..ctor(String assemblyName)
        /_/src/libraries/System.Reflection/tests/AssemblyNameTests.cs(640,0): at System.Reflection.Tests.AssemblyNameTests.ToStringTest(String name)
  Finished:    System.Reflection.Tests
```

BartoszKlonowski commented:
@jkotas Sure, no problem!

BartoszKlonowski committed:
Test the AssemblyName.ToString method for empty Name

The empty name of AssemblyName needs to be tested separately so that the
ToStringTest will stay unmodified and the test scenario won't change,
but at the same time the empty string can be passed without using the
constructor.
Note that the AssemblyName constructor throws the exception when an empty
string is given as the name.

jkotas commented on src/libraries/System.Reflection/tests/AssemblyNameTests.cs:
```diff
@@ -407,7 +408,7 @@ public void Name_Set_FullName_Invalid()
         {
```
This test is testing the exact same as the last theory of `Name_Set_FullName` . It should be deleted instead.

jkotas commented on src/libraries/System.Reflection/tests/AssemblyNameTests.cs:
```diff
@@ -640,6 +641,16 @@ public void ToStringTest(string name)
         }
 
+        [Theory]
+        [InlineData("")]
```
This can be Fact instead since it has just one `InlineData` and the test name suggests that it is only testing EmptyName.

BartoszKlonowski committed:
Simplify the unit test for empty AssemblyName.Name

The simplification includes:
1. Removing the redundant test for empty FullName as it now checks the
   same case as the Name_Set_FullName Theory
2. Make the ToStringEmptyNameTest a Fact instead of Theory as there's
   only one case which can be done as an InlineData.
